src/Symfony/Component/Messenger/Transport/Doctrine/DoctrineReceiver.php

```diff
@@ -0,0 +1,127 @@
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class DoctrineReceiver implements ReceiverInterface
+{
+    private $connection;
+    private $serializer;
+    private $shouldStop = false;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer = null)
+    {
+        $this->connection = $connection;
+        $this->serializer = $serializer ?? new PhpSerializer();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(callable $handler): void
+    {
+        while (!$this->shouldStop) {
+            try {
+                $doctrineEnvelope = $this->connection->get();
+            } catch (DBALException $exception) {
+                throw new TransportException($exception->getMessage(), 0, $exception);
+            }
+
+            if (null === $doctrineEnvelope) {
+                $handler(null);
+
+                usleep($this->connection->getConfiguration()['loop_sleep'] ?? 200000);
+
+                continue;
+            }
+
+            try {
+                $envelope = $this->serializer->decode([
+                    'body' => $doctrineEnvelope['body'],
+                    'headers' => $doctrineEnvelope['headers'],
+                ]);
+            } catch (MessageDecodingFailedException $exception) {
+                $this->rejectDoctrineMessage($doctrineEnvelope['id']);
+
+                throw $exception;
+            }
+
+            $envelope = $envelope->with(new DoctrineReceivedStamp($doctrineEnvelope['id']));
+            $handler($envelope);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function ack(Envelope $envelope): void
+    {
+        $this->ackDoctrineMessage($this->findDoctrineReceivedStamp($envelope)->getId());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reject(Envelope $envelope): void
+    {
+        $this->rejectDoctrineMessage($this->findDoctrineReceivedStamp($envelope)->getId());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stop(): void
+    {
+        $this->shouldStop = true;
+    }
+
+    private function findDoctrineReceivedStamp(Envelope $envelope): DoctrineReceivedStamp
+    {
+        /** @var DoctrineReceivedStamp|null $doctrineReceivedStamp */
+        $doctrineReceivedStamp = $envelope->last(DoctrineReceivedStamp::class);
+
+        if (null === $doctrineReceivedStamp) {
+            throw new LogicException('No DoctrineReceivedStamp found on the Envelope.');
+        }
+
+        return $doctrineReceivedStamp;
+    }
+
+    private function ackDoctrineMessage(string $id)
+    {
+        try {
+            $this->connection->ack($id);
+        } catch (DBALException $exception) {
+            throw new TransportException($exception->getMessage(), 0, $exception);
```

Comment:
I'd do this catch in the `Connection` so you don't need these two `ackDoctrineMessage` and `rejectDoctrineMessage` methods.